hbase-client/src/main/java/org/apache/hadoop/hbase/security/NettyHBaseSaslRpcClientHandler.java

```diff
@@ -114,7 +122,10 @@ public void handlerAdded(ChannelHandlerContext ctx) {
       if (initialResponse != null) {
         writeResponse(ctx, initialResponse);
       }
-      tryComplete(ctx);
+      // HBASE-23881 We do not want to check if the SaslClient thinks the handshake is
```

Comment:
Is this correct? 

The comment is
```
      * Determines whether the authentication exchange has completed.
      * This method may be called at any time, but typically, it
      * will not be called until the caller has received indication
      * from the server
      * (in a protocol-specific manner) that the exchange has completed.
```

It does not say that we 'must' call it after we receive the reponse from server.

And I'm curious that, what is the sasl client implementation? It just marks itself as complete succeeded, without any server response, and then after the server response it could revert itself as auth failed? I think this is a broken sasl client implementation?

Reply:
> It does not say that we 'must' call it after we receive the reponse from server.

I was initially thinking the same thing: "why the heck does this implementation say 'i'm done'??". I can read it both ways -- reading the JDK-provided implementations, the "you should wait for a response from the server" was the only interpretation which made any sense to me.

> what is the sasl client implementation

I looked at what was provided in openjdk8 because I had the same question about which mechanisms actually acted this way :). Surveying all mechanisms included, we have two categories. One category is that the mechanism's impl for `isComplete()` returns true after a single call to `evaluateChallenge`, and the other category is that it takes multiple calls to evaluateChallenge to eventually have `isComplete` return true. You can also think of these as "challenge is exactly one step" and "challenge is multiple steps", respectively

* Plain - one step
* DigestMD5 - multiple steps
* NTLM - multiple steps
* GSSAPI/KRB5 -- multiple steps
* CramMD5 - one step
* External - one step

This test is using Plain which is in the "one step" category, while all over sasl-based mechanisms previously in HBase are in the "multiple steps" bucket :).

I completely understand your initial reaction that this mechanism is "broken", but, after doing the digging I have, I believe our interpretation of Sasl was just unaware of how some other implementations work in this regard.

Reply:
Any chance the second part of your comment be added as a code comment? I was digging into the patch and this comment made it much clear for me. That also explains why we don't see this issue in the GSS/Kerberos implementation.

Reply:
Sure, let me add that, too.